impl/src/main/java/org/jboss/weld/executor/CommonForkJoinPoolExecutorServices.java

```diff
@@ -42,4 +49,42 @@ public void cleanup() {
     protected int getThreadPoolSize() {
         return ForkJoinPool.getCommonPoolParallelism();
     }
+
+    /**
+     * If in SE environment, this method wraps the collection of tasks in such a way that it sets the TCCL to null before
+     * executing them and re-sets it to previous version once done. This affect the CL Weld will pick up when invoking
+     * {@link org.jboss.weld.environment.deployment.WeldResourceLoader#getClassLoader()}.
+     *
+     * @see WELD-2494
+     */
+    @Override
+    public <T> Collection<? extends Callable<T>> wrap(Collection<? extends Callable<T>> tasks) {
+        // try to detect environment, if it's SE, we want to null TCCL, otherwise leave it as it is
+        if (isSeEnvironment()) {
+            List<Callable<T>> wrapped = new ArrayList<>(tasks.size());
+            for (Callable<T> task : tasks) {
+                wrapped.add(() -> {
+                    ClassLoader oldTccl = Thread.currentThread().getContextClassLoader();
+                    Thread.currentThread().setContextClassLoader(null);
+                    try {
+                        return task.call();
+                    } finally {
+                        Thread.currentThread().setContextClassLoader(oldTccl);
+                    }
+                });
+            }
+            return wrapped;
+        } else {
+            return tasks;
+        }
+    }
+
+    /**
+     * Tries to detect SE environment by loadability of SE-specific class. If successful, returns true; false otherwise.
+     *
+     * @return Environments.SE if {@link org.jboss.weld.environment.se.Weld} is loadable, null otherwise
+     */
+    private boolean isSeEnvironment() {
+        return Reflections.isClassLoadable("org.jboss.weld.environment.se.Weld", WeldClassLoaderResourceLoader.INSTANCE);
```

Comment:
There is already `org.jboss.weld.interceptor.util.InterceptionUtils.detectEnvironment()`. We should probably extract this logic into a util class. Or find a better way, e.g. enhance the `org.jboss.weld.Container`.

Reply:
I was thinking of extracting it but `InterceptionUtils.detectEnvironment()` cares about SE and servlet, while here we only care about SE (e.g. this method is way simpler).
Not to mention this "hacky" approach probably shouldn't be publicly exposed as common util method.

Reply:
I think it's better to "expose" one hacky util method than to have two hidden almost identical hacky methods. Util classes are not supposed to be public API of Weld anyway.

Also we could try to improve the dection logic so that it's not so hacky...